daemon/logger/loggerutils/logfile.go

```diff
@@ -179,45 +179,54 @@ func (w *LogFile) WriteLogEntry(msg *logger.Message) error {
 	return err
 }
 
-func (w *LogFile) checkCapacityAndRotate() error {
+func (w *LogFile) checkCapacityAndRotate() (retErr error) {
 	if w.capacity == -1 {
 		return nil
 	}
 
-	if w.currentSize >= w.capacity {
-		w.rotateMu.Lock()
-		fname := w.f.Name()
-		if err := w.f.Close(); err != nil {
-			// if there was an error during a prior rotate, the file could already be closed
-			if !errors.Is(err, os.ErrClosed) {
-				w.rotateMu.Unlock()
-				return errors.Wrap(err, "error closing file")
-			}
-		}
-		if err := rotate(fname, w.maxFiles, w.compress); err != nil {
-			w.rotateMu.Unlock()
-			return err
-		}
-		file, err := openFile(fname, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
-		if err != nil {
+	if w.currentSize < w.capacity {
+		return nil
+	}
+
+	w.rotateMu.Lock()
+	defer func() {
+		if retErr != nil || w.maxFiles <= 1 || !w.compress {
 			w.rotateMu.Unlock()
-			return err
 		}
-		w.f = file
-		w.currentSize = 0
-		w.notifyRotate.Publish(struct{}{})
+	}()
 
-		if w.maxFiles <= 1 || !w.compress {
-			w.rotateMu.Unlock()
-			return nil
+	fname := w.f.Name()
+	if err := w.f.Close(); err != nil {
+		// if there was an error during a prior rotate, the file could already be closed
+		if !errors.Is(err, os.ErrClosed) {
+			return errors.Wrap(err, "error closing file")
 		}
+	}
 
-		go func() {
-			compressFile(fname+".1", w.lastTimestamp)
-			w.rotateMu.Unlock()
-		}()
+	if err := rotate(fname, w.maxFiles, w.compress); err != nil {
+		return err
+	}
+	file, err := openFile(fname, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
+	if err != nil {
+		return err
+	}
+	w.f = file
+	w.currentSize = 0
+	w.notifyRotate.Publish(struct{}{})
+
+	if w.maxFiles <= 1 || !w.compress {
```

Comment:
Perhaps `w.rotateMu.Unlock()` here, so that we don't have to do the extra check in the `defer`?

Reply:
Updated, but more as we discussed on last weeks call.
This now sets a var for `noCompress := i w.maxFiles <= 1 || !w.compress`, then we check that var both in the defer and here.

Additionally added some docs around that in the defer.